### ISSUE TITLE
Missing Narrowing for class defined with a decorator applied

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#5579]
**Describe the bug**
```python
from collections.abc import Iterable
from typing import Callable, TypeVar, Union

T = TypeVar("T")


def class_dec() -> Callable[[type[T]], type[T]]:
    ...


@class_dec()
class Foo:
    def get_features(self) -> list[str]:
        ...


def process(specs: Union[Iterable[str], Foo]) -> None:
    if isinstance(specs, Foo):
        features = specs.get_features()
    else:
        features = specs

   # Error here Argument of type "Iterable[str] | Foo" cannot be assigned to 
   # parameter "__iterable" of type "Iterable[_T@set]" in function "__init__"
    print(set(features)) 
```

If class decorator is removed then this code type checks. But with class decorator, the narrowing no longer holds.

**VS Code extension or command-line**
pyright 1.1.318. pyright 1.1.317 and before passes type checking.

**Comment by erictraut** [comment]
This is by design, although I can understand why it would be somewhat unintuitive.

Without the class decorator, the symbol `Foo` (when used as an argument to `isinstance`) refers precisely to the class `Foo` and no other types. When you apply the class decorator, the result is type `type[Foo]` which refers to "the set of types that contains the type `type[Foo]` and all subtypes thereof".  When you use this value as the second argument to `isinstance`, it is not safe for pyright to assume that `Foo` can be eliminated from the union when narrowing in the negative ("else") case.

It's comparable to the following code:

```python
def func(specs: Iterable[str] | Foo, cls: type[Foo]) -> None:
    if isinstance(specs, cls):
        reveal_type(specs) # Foo
    else:
        reveal_type(specs) # Iterable[str] | Foo
```

Mypy and pyright produce the same type narrowing for my code snippet above. Mypy doesn't produce the same results for your original code, but that's because mypy does not honor class decorators; it simply ignores them. If mypy were to implement support for class decorators, it would produces the same result as pyright.

**Comment by hmc-cs-mdrissi** [comment]
That seems inconsistent in that a variation of class decorator,

```python
def class_dec(cls: T) -> T:
  ...

@class_dec
class Foo:
  ...
```

A similar argument feels like it should apply here. The decorator takes in type[Foo] and it would in theory be valid for it to return subtype. But for T -> T case narrowing continues to be allowed, while type[T] -> type[T] case narrowing is hole?

I can even go one step closer and define class_dec like,

```python
T = TypeVar("T", bound=type)

def class_dec(cls: T) -> T:
  ...

@class_dec
class Foo:
  ...
```

and narrowing is allowed here. I'm unsure I see much difference in type[T] -> type[T] for unbound type var vs T -> T where T is bound to type.

I'll probably just adjust my code to one of variations that works.

**Comment by erictraut** [comment]
I didn't notice that this was a new behavior. I'm going to reopen and investigate further.

**Comment by AlexanderPodorov** [comment]
@AsiPanda , @erictraut ,
Just got the same warning after updating `pylance` from `v2023.7.30` to `v2023.7.40`.
The code below produces no warnings in  `v2023.7.30` in `strict` mode, but shows the warning in `v2023.7.40` in `strict` mode.
I was able to work out the MRE based on SQLAlchemy use case:
```python
from typing import Generic, TypeVar

_T = TypeVar("_T")


def class_logger(cls: type[_T]) -> type[_T]:
    return cls


@class_logger
class Mapper(
    Generic[_T],
):
    pass


t = Mapper[int]  # Type "Mapper[Unknown]" is already specialized

# No warnings in v2023.7.30 (strict mode)
# Shows above warning in v2023.7.40 (strict mode)

```

**Comment by erictraut** [comment]
Some background... 
The type annotation `type[X]` in the Python type system represents "the set containing the class X and all subtypes of class X". For example, the type annotation `type[int]` refers to "the class `int` and any subtypes of `int`". The expression `int` refers to the class `int`, and its type is `type[int]`. However, in this case, we know that it represents "only the class `int`, not subtypes of class `int`". This distinction is important in certain cases, such as for `isinstance` type narrowing. For this reason, pyright internally differentiates between the two variations of `type[int]` — one that represents a set of classes and the other that represents a specific class but not its subtypes. 

With that background in mind, now let's consider what the type annotation `type[T]` means. Let's assume that the constraint solver determines that `T` is `int`, so `type[T]` becomes `type[int]`. But which of the two variants of `type[int]` does this represent — the one that includes subtypes of `int` or the one that doesn't? There's nothing in the type spec that indicates the right answer, so this is left up to each type checker to decide. Pyright contains various heuristics to make this determination.

Another thing that is affected by this distinction is whether a generic class (like `dict`) can remain unspecialized or whether it should be specialized with default type arguments (typically `Unknown`). 

Consider the following three cases:
```python
def func1(x: type[T]) -> T: ...
reveal_type(func1(dict)) # `dict[_KT, _VT]` or `dict[Unknown, Unknown]`?

def func2(x: type[T]) -> type[T]: ...
reveal_type(func2(dict)) # `type[dict[_KT, _VT]]` or `type[dict[Unknown, Unknown]]`?

def func3(x: T) -> T: ...
reveal_type(func3(dict)) # `type[dict[_KT, _VT]]` or `type[dict[Unknown, Unknown]]`?
```

In pyright 1.1.319, `func1` and `func2` are filled in with `Unknown` type arguments but `func3` is not. This is consistent with mypy's behavior. But the behavior of `func2` creates problems for class decorators. Mypy doesn't have this problem because it ignores class decorators.

I've decided to change pyright's behavior for `func2` so it acts like `func3`. This deviates from mypy, but I think there's good justification.

This change fixes the class decorator regressions reported above — including the unexpected type narrowing behavior reported by @hmc-cs-mdrissi.

This change will be included in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.320, which I just published. It will also be included in a future version of pylance.
